osmoutils/ibc.go

```diff
@@ -3,11 +3,25 @@ package osmoutils
 import (
 	"encoding/json"
 
-	transfertypes "github.com/cosmos/ibc-go/v3/modules/apps/transfer/types"
-	channeltypes "github.com/cosmos/ibc-go/v3/modules/core/04-channel/types"
-	ibcexported "github.com/cosmos/ibc-go/v3/modules/core/exported"
+	transfertypes "github.com/cosmos/ibc-go/v4/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v4/modules/core/04-channel/types"
+	ibcexported "github.com/cosmos/ibc-go/v4/modules/core/exported"
 )
 
+// NewStringErrorAcknowledgement returns a new instance of Acknowledgement using an Acknowledgement_Error
+// type in the Response field. These errors differ from the IBC errors in that we use custom error strings.
+// NOTE: Acknowledgements are written into state and thus, changes made to error strings included in packet acknowledgements
+// risk an app hash divergence when nodes in a network are running different patch versions of software.
+func NewStringErrorAcknowledgement(err string) channeltypes.Acknowledgement {
+	// ToDo: Do we want to do this or do we want to use the IBC errors and emit the string?
```

Comment:
I don;t understand the situation enough to have any opinions here

Reply:
is there any particular reason you guys have chosen to use this in favour of [channeltypes.NewErrorAcknowledgement](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/types/acknowledgement.go#L32)? 

Acknowledgements are written into the ibc store on `RecvPacket` so that the counterparty can verify their existence on `AcknowledgePacket`, and delete commitments (and in the case of ordered channels, increment an expected sequence).

Suppose an error string which is included in a packet acknowledgment is modified in a patch version of an osmosis release. Nodes running lets say `v14.0.0` and those running `v14.0.1` will produce different app hashes.

We made this change to `NewErrorAcknowledgement` in ibc-go/v4 as we deemed it unsafe as developers may unknowingly change a string value which could lead to the above. Also, to note, error values returned from protobuf codecs `UnmarshalJSON` produce non-deterministic output. So for example, when unmarshaling packet data `[]byte` to a well known type in `OnRecvPacket` its not safe to include the error in the ack. See for example we [omit the error in transfer acks](https://github.com/cosmos/ibc-go/blob/main/modules/apps/transfer/ibc_module.go#L176-L179). I think you guys are using `encoding/json` in places here tho, which I hope is a bit more sane than the proto shenanigans.

Reply:
☝️ @ValarDragon This is the reason. I think it makes sense to change it, but I didn't want to do this without a discussion because it changes the behaviour (errors in ack vs errors in events).

I think a middle way is a good option, where we can add some information in the acks, but only from this packet and not from underlying errors.

I don't se why unmarshalling the error would lead to non-determinism as it's just a string (though I'm sure I'm wrong)

Reply:
IIRC its because the `jsonpb` unmarshaling [uses maps](https://github.com/golang/protobuf/blob/master/jsonpb/decode.go#L290-L303). Ranging on these maps in the deserialisation process will be non deterministic and therefore, you may get a different key returned in the error msg for a particular struct value.

Reply:
that makes sense. And I think that aligns with only using string errors if they do not rely on errors from other parts of the stack. 

I like having data in acks (errors in this case, but the success could also have data) because it allows us to do communication in cases where events aren't available (cosmwasm). I think the Async-ICQ implementation does something similar. But you're right about the risk of halts if we're not careful when changing these. 

I should refactor the code so that it's very hard to accidentally introduce changes to the merkelized errors in minor versions. This change will probably need to happen both here and in the rate limit middleware/contract